src/flavors/palouse/config.json

```diff
@@ -49,6 +49,3265 @@
+                "grassed_waterways": "yes",
+                "weed_control": "yes",
+                "tree_shrub": "yes",
+                "upland_habitat": "yes",
+                "hedgerow_planting": "yes",
+                "open_channel": "yes",
+                "fencing": "yes"
+              }
+            }
+          ],
+          "layout": {
+            "start": 1,
+            "end": 4
+          }
+        },
+        {
+          "header": "Reporting Activity",
+          "type": "lineChart",
+          "xAxisLabel": "Date",
+          "yAxisLabel": "No. VSP reports",
+          "layout": {
+            "start": 4,
+            "end": 13,
+            "height": 300
+          }
+        },
+        {
+          "header": "Overview by Stewardship Category",
+          "type": "fixedTable",
+          "stripeColor": "#f8f5f5",
+          "layout": {
+            "start": 1,
+            "end": 13,
+            "height": 500
+          },
+          "widgetStateControls": [
+            {
+              "defaultValue": "all-dates",
+              "name": "vsp-summary-date-range",
+              "title": "Date range",
+              "options": [
+                {
+                  "value": "all-dates",
+                  "label": "All dates"
+                },
+                {
+                  "value": "pre-2011",
+                  "label": "Pre-2011"
+                }
+              ]
+            }
+          ],
+          "columns": [
+            {
+              "header": "Stewardship category",
+              "type": "string",
+              "fractionalWidth": 3
+            },
+            {
+              "header": "Countywide",
+              "type": "numeric"
+            },
+            {
+              "header": "Palouse WAU",
+              "type": "numeric",
+              "filter": ["==", ["get-val", "analysis_unit"], "palouse"]
+            },
+            {
+              "header": "Snake WAU",
+              "type": "numeric",
+              "filter": ["==", ["get-val", "analysis_unit"], "snake"]
+            },
+            {
+              "header": "Rock Lake WAU",
+              "type": "numeric",
+              "filter": ["==", ["get-val", "analysis_unit"], "rock"]
+            },
+            {
+              "header": "2021 objective",
+              "type": "numeric"
+            },
+            {
+              "header": "2021 status",
+              "type": "badge"
+            },
+            {
+              "header": "2021 enhanced objective",
+              "type": "numeric"
+            },
+            {
+              "header": "2021 enhanced status",
+              "type": "badge"
+            }
+          ],
+          "rows": [
+            {
+              "cells": [
+                {
+                  "value": "Range Management",
+                  "label": "(Acreage Totals)"
+                },
+                {
+                  "value": [
```

Comment:
@goldpbear What is the schema of "value", and how would we model it in a database?

Reply:
`value` here is a `ValueExpression`, detailed in Quip.

I'm not sure we should model value expressions like this. Can they just be a JSON string? Modeling them will be difficult, I think.

Reply:
>Can they just be a JSON string? 

Yes, Postgres supports JSON. But we'd be going NoSQL, which has some tradeoffs that I'm not too familiar. For example, migrating and querying NoSQL data can be very difficult.

Still, this might be a valid use case for NoSQL - I don't know. Sorry that I didn't check out that Quip doc earlier. It would have been nice to research what a NoSQL db entails, before we started the implementation. It's a significant architectural decision that I'm not familiar with.

I think there might have been a simpler, but less powerful, schema we could have used that would preserve relational data. At this point, however, I think it'll be better to research what a NoSQL approach entails, since we've already stepped in that direction...

Reply:
I was just imagining these value expressions would live as text in the database, similar (I think) to how a Place's data JSON blob is currently stored.

Because really they're just a piece of text content, I suppose, similar to a paragraph. As you author a dashboard you author these value expressions to express quantities, etc. Maybe you make a mistake entering a value expression, but you would test the results and edit it accordingly.

Reply:
>I was just imagining these value expressions would live as text in the database, similar (I think) to how a Place's data JSON blob is currently stored.

There are numerous tradeoffs to that approach, and it's a big topic - if you search "sql vs nosql", you will see that it's not so simple.

For example, what if we needed to change the name of our widget's field `vsp-summary-date-range` to something else? Or delete it? On a relational db, it's trivial using SQL constraints. But in NoSQL, we'd have to perform a "migration" to find/replace all `vsp-summary-date-range` strings - basically a custom script.

>As you author a dashboard you author these value expressions to express quantities...

Authoring content in JSON is programmer-friendly, but not friendly to most users. If we wanted to build a GUI on top of this, we'd have to validate and serialize the models from the UI to match the powerful/complex JSON that underlies it. Whereas a relational model would allow us to rely on the DB and ORM to validate the input data for us.

---

All in all, I wouldn't worry about refactoring or anything. I think we'll stick with the current NoSQL for now, until we find clear evidence otherwise.

Reply:
Yeah, if we did need to perform a database-wide update like you describe there wouldn't be a ready way to do it. But I also wonder how often that use case will come up.

> Authoring content in JSON is programmer-friendly, but not friendly to most users and non-programmers.

That's definitely true, but I wasn't imagining users would author this kind of content. I guess I just implicitly assumed that complex dashboard widgets which require these sorts of value expressions would be authored by us, for custom applications.

I might be wrong, but building a GUI capable of configuring the VSP dashboards we've designed is approaching a level of complexity similar to that of Mapbox Studio, and that feels like a long way to reach. Is that a direction we want to take the product? We should probably have an idea of whether that's true.

But maybe a GUI one day could have a library of "basic" dashboard widgets which come pre-configured, and users could select from among them.

Reply:
Though actually, I guess we could model expressions relationally, maybe with less trouble than I was originally thinking.

Each expression has a type, which is the first element of the array, and then it has a certain number of arguments, which can either be pointers to other sub-expression arrays, or a literal value (a string, number, or boolean).

So constructing an expression would entail following all the pointers to all the sub-arrays and serializing that as an array of arrays.

Though I'm still not sure it's worth it? But I'm also not very qualified to make a call on that...

Reply:
>Yeah, if we did need to perform a database-wide update like you describe there wouldn't be a ready way to do it. But I also wonder how often that use case will come up.

It's one thing to build a feature that's YAGNI, but in this case, going NoSQL is an architectural decision that will have long term effects on how our dashboards will be modelled. I'm concerned that are committing to dashboards that can only be authored/edited via a JSON interface.

>That's definitely true, but I wasn't imagining users would author this kind of content.

Any non-technical users will have difficulty editing raw JSON. This includes customer service reps, community engagement consultants, etc. Maybe I'm missing something, but requiring JSON basically limits the authoring to engineers only.

>So constructing an expression would entail following all the pointers to all the sub-arrays and serializing that as an array of arrays.

I imagine that's how Mapbox Studio is built. It's biting off a lot, and it would take great effort to model all of that out. That's why I would have explored a simpler, less powerful, relational data structure.

Still, I don't know which approach is right. I would have to think on it. I just wanted to raise awareness about the ramifications of our current data structure.

---

I think the pain points that I'm raising would be clearer if we had a habit of building back end data models when shipping a feature. NoSQL has it's uses, but SQL is the tried and true data store for web apps. So I just want to emphasize that we should be careful when we veer off the beaten path.

Reply:
> It's one thing to build a feature that's YAGNI, but in this case, going NoSQL is an architectural decision that will have long term effects on how our dashboards will be modelled.

I didn't mean to suggest we use a NoSQL architecture, I just meant we treat the value expressions as un-modeled text content. Or, we figure out how to model expressions relationally.

> Any non-technical users will have difficulty editing raw JSON. This includes customer service reps, community engagement consultants, etc.

I've seen cases where semi-technical folks author Mapbox style expressions, which is part of the reason I thought it would be ok to take that approach here. Still, I don't know what the process of learning to use them would be like.

Reply:
>I didn't mean to suggest we use a NoSQL architecture, I just meant we treat the value expressions as un-modeled text content. Or, we figure out how to model expressions relationally.

I'll have to think more about what it will take to model this relationally. But from the looks of it, it's not very SQL-friendly. Of course, it's not impossible, but it would be fairly complicated. The concept of "pointers to other sub-expression arrays, or a literal value (a string, number, or boolean, or object)" isn't simple to do in a relational db. 

I think the models would still be fairly complex, so I would have to think about whether it would be better off going NoSQL.

Reply:
One thing that occurred to me, which might make sense to iron out now, is how to handle the serialization of `ValueExpression`s. One of the reasons the configuration for the VSP tables is so large is that many multi-line `ValueExpression`s are repeated across numerous table cells.

For example, we might have a table tow configuration like:
```
{
  "cells": [
    {
      "value": [...]
    },
    {
      "value": [...]
    }
  ]
}
```
where each `[...]` `ValueExpression` is identical. (Note that the evaluated values of these expressions might be different, because table columns can impose filters that run before cell values are computed).

All these repeated expressions add up. It's particularly bad in the `garfield` config.

So would it be better to serialize all `ValueExpressions` in their own little section, and then use numerical pointers in the cells to refer to `ValueExpression`s? My guess is this would cut out multiple hundreds of lines.

Reply:
For now, I would advise against doing this.

If we did do it, we'd want to refer to *all* of our `ValueExpression`s in this relational manner. But having some of our `ValueExpression`'s as literals, while others are stored in a separate "table", is only going to complicate the schema, and complicate whatever code has to parse this data structure as well.

Reply:
> If we did do it, we'd want to refer to all of our ValueExpressions in this relational manner. 

That makes sense. Yeah, I wouldn't want to deal with a structure where some expressions are references to a separate object, and others are not.

I'll leave the schema as is.